runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -898,3 +901,600 @@ JITServerLocalSCCAOTDeserializer::updateSCCOffsets(SerializedAOTMethod *method,
+   _classChainMap(decltype(_classChainMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _wellKnownClassesMap(decltype(_wellKnownClassesMap)::allocator_type(TR::Compiler->persistentAllocator()))
+   { }
+
+
+void
+JITServerNoSCCAOTDeserializer::invalidateClassLoader(J9VMThread *vmThread, J9ClassLoader *loader)
+   {
+   assertExclusiveVmAccess(vmThread);
+
+   auto p_it = _classLoaderPtrMap.find(loader);
+   if (p_it == _classLoaderPtrMap.end())// Not cached
+      return;
+
+   uintptr_t id = p_it->second;
+   auto i_it = _classLoaderIdMap.find(id);
+   TR_ASSERT(i_it != _classLoaderIdMap.end(), "Broken two-way map");
+
+   // Mark entry as unloaded
+   i_it->second = NULL;
+   _classLoaderPtrMap.erase(p_it);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Invalidated class loader %p ID %zu", loader, id);
+   }
+
+void
+JITServerNoSCCAOTDeserializer::invalidateClass(J9VMThread *vmThread, J9Class *ramClass)
+   {
+   assertExclusiveVmAccess(vmThread);
+
+   auto p_it = _classPtrMap.find(ramClass);
+   if (p_it == _classPtrMap.end()) // Not cached or already marked invalid
+      return;
+
+   uintptr_t id = p_it->second;
+   auto i_it = _classIdMap.find(id);
+   TR_ASSERT(i_it != _classIdMap.end(), "Broken two-way map");
+
+   if (i_it->second._ramClass)
+      {
+      // Class ID is valid. Keep the entry, but mark it as unloaded.
+      i_it->second._ramClass = NULL;
+      }
+   else
+      {
+      // Class ID is invalid. Remove the entry so that it can be replaced
+      // if a matching version of the class is ever loaded in the future.
+      _classIdMap.erase(i_it);
+      }
+
+   _classPtrMap.erase(p_it);
+
+   // Invalidate any methods that might have this RAM class recorded as
+   // their defining class.
+   for (uint32_t i = 0; i < ramClass->romClass->romMethodCount; i++)
+      invalidateMethod(&ramClass->ramMethods[i]);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Invalidated RAMClass %p ID %zu", ramClass, id);
+   }
+
+void
+JITServerNoSCCAOTDeserializer::invalidateMethod(J9Method *method)
+   {
+   auto p_it = _methodPtrMap.find(method);
+   if (p_it == _methodPtrMap.end()) // Not cached or already marked as invalid
+      return;
+
+   uintptr_t id = p_it->second;
+   auto i_it = _methodIdMap.find(id);
+   TR_ASSERT(i_it != _methodIdMap.end(), "Broken two-way map");
+
+   i_it->second = NULL;
+
+   _methodPtrMap.erase(p_it);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Invalidated RAMMethod %p ID %zu", method, id);
+   }
+
+// Free all the values (which must have been allocated with TR::Compiler->persistentGlobalMemory()->allocatePersistentMemory)
+// to the given map. Note that this leaves invalid pointers in the map, so the map should be cleared after this operation.
```

Comment:
Is there a reason the map can't be cleared in this function itself?

Reply:
It can be. Maybe having a separate function isn't worth it in this case. Changed.